### PR TITLE
[Brave News]: Add `Following` feed

### PR DESCRIPTION
**Pull request by fallaciousreasoning**
Resolves https://github.com/brave/brave-browser/issues/34143

In addition:
- Use Block ranking for Channel feeds
- Don't include Discover articles in Channel blocks
- Fixes for Link click handling (enabling Control/Command/Middle click for open in new tab)

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_news/browser/feed_v2_builder.cc [hunk 1]
```diff
@@ -639,14 +680,27 @@ void FeedV2Builder::BuildChannelFeed(const std::string& channel,
...
             }
 
-            base::ranges::sort(result->items, [](const auto& a, const auto& b) {
-              return GetPopRecency(b->get_article()->data) <
-                     GetPopRecency(a->get_article()->data);
-            });
+            auto articles = GetArticleInfos(locale, feed_items, publishers,
+                                            builder->signals_);
+
+            constexpr size_t kBlocksPerAd = 2;
```
Comment (fallaciousreasoning) [inline components/brave_news/browser/feed_v2_builder.cc:689]: Note: We might have different ad ratios in the different feeds - not sure yet.

**Comment by simonhong** [comment]
I got a below failure after startup with this PR on macOS.
```
0   libbase.dylib                       0x00000001108f0ee2 base::debug::CollectStackTrace(void const**, unsigned long) + 18
1   libbase.dylib                       0x00000001108d3a83 base::debug::StackTrace::StackTrace() + 19
2   libbase.dylib                       0x00000001107bfe35 logging::LogMessage::~LogMessage() + 181
3   libchrome_dll.dylib                 0x000000011d30b049 mojo::internal::Serializer<brave_news::mojom::FeedV2DataView, mojo::StructPtr<brave_news::mojom::FeedV2>>::Serialize(mojo::StructPtr<brave_news::mojom::FeedV2>&, mojo::internal::MessageFragment<brave_news::mojom::internal::FeedV2_Data>&) + 777
4   libchrome_dll.dylib                 0x000000011d2f3df2 brave_news::mojom::BraveNewsController_GetFeedV2_ProxyToResponder::Run(mojo::StructPtr<brave_news::mojom::FeedV2>) + 178
5   libchrome_dll.dylib                 0x000000011d2fc0c6 base::internal::Invoker<base::internal::BindState<void (brave_news::mojom::BraveNewsController_GetFollowingFeed_ProxyToResponder::*)(mojo::StructPtr<brave_news::mojom::FeedV2>), std::__Cr::unique_ptr<brave_news::mojom::BraveNewsController_GetFollowingFeed_ProxyToResponder, std::__Cr::default_delete<brave_news::mojom::BraveNewsController_GetFollowingFeed_ProxyToResponder>>>, void (mojo::StructPtr<brave_news::mojom::FeedV2>)>::RunOnce(base::internal::BindStateBase*, mojo::StructPtr<brave_news::mojom::FeedV2>&&) + 54
6   libchrome_dll.dylib                 0x000000011d2f303c base::OnceCallback<void (mojo::StructPtr<brave_news::mojom::FeedV2>)>::Run(mojo::StructPtr<brave_news::mojom::FeedV2>) && + 76
7   libchrome_dll.dylib                 0x000000012036752d brave_news::BraveNewsController::GetFeedV2(base::OnceCallback<void (mojo::StructPtr<brave_news::mojom::FeedV2>)>) + 93
8   libchrome_dll.dylib                 0x000000011d2f8c1b brave_news::mojom::BraveNewsControllerStubDispatch::AcceptWithResponder(brave_news::mojom::BraveNewsController*, mojo::Message*, std::__Cr::unique_ptr<mojo::MessageReceiverWithStatus, std::__Cr::default_delete<mojo::MessageReceiverWithStatus>>) + 1739
9   libmojo_public_cpp_bindings.dylib   0x00000001104b5196 mojo::InterfaceEndpointClient::HandleValidatedMessage(mojo::Message*) + 1286
10  libmojo_public_cpp_bindings.dylib   0x00000001104bd623 mojo::MessageDispatcher::Accept(mojo::Message*) + 131
11  libmojo_public_cpp_bindings.dylib   0x00000001104b72e2 mojo::InterfaceEndpointClient::HandleIncomingMessage(mojo::Message*) + 114
12  libmojo_public_cpp_bindings.dylib   0x00000001104c2011 mojo::internal::MultiplexRouter::ProcessIncomingMessage(mojo::internal::MultiplexRouter::MessageWrapper*, mojo::internal::MultiplexRouter::ClientCallBehavior, base::SequencedTaskRunner*) + 1057
13  libmojo_public_cpp_bindings.dylib   0x00000001104c15fe mojo::internal::MultiplexRouter::Accept(mojo::Message*) + 414
14  libmojo_public_cpp_bindings.dylib   0x00000001104bd661 mojo::MessageDispatcher::Accept(mojo::Message*) + 193
15  libmojo_public_cpp_bindings.dylib   0x00000001104ac52a mojo::Connector::DispatchMessage(mojo::ScopedHandleBase<mojo::MessageHandle>) + 634
16  libmojo_public_cpp_bindings.dylib   0x00000001104ad22e mojo::Connector::ReadAllAvailableMessages() + 286
17  libmojo_public_cpp_bindings.dylib   0x00000001104acf49 mojo::Connector::OnWatcherHandleReady(char const*, unsigned int) + 73
18  libmojo_public_cpp_bindings.dylib   0x00000001104ad9b2 base::RepeatingCallback<void (unsigned int)>::Run(unsigned int) const & + 82
19  libmojo_public_system_cpp.dylib     0x000000011042eddc base::RepeatingCallback<void (unsigned int, mojo::HandleSignalsState const&)>::Run(unsigned int, mojo::HandleSignalsState const&) const & + 92
20  libmojo_public_system_cpp.dylib     0x000000011042eb98 mojo::SimpleWatcher::OnHandleReady(int, unsigned int, mojo::HandleSignalsState const&) + 264
21  libmojo_public_system_cpp.dylib     0x000000011042f1d5 base::internal::Invoker<base::internal::BindState<void (mojo::SimpleWatcher::*)(int, unsigned int, mojo::HandleSignalsState const&), base::WeakPtr<mojo::SimpleWatcher>, int, unsigned int, mojo::HandleSignalsState>, void ()>::RunOnce(base::internal::BindStateBase*) + 101
22  libbase.dylib                       0x000000011079a1a4 base::OnceCallback<void ()>::Run() && + 68
23  libbase.dylib                       0x00000001108457c7 base::TaskAnnotator::RunTaskImpl(base::PendingTask&) + 263
24  libbase.dylib                       0x0000000110875db7 base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWorkImpl(base::LazyNow*) + 1655
25  libbase.dylib                       0x00000001108753c4 base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWork() + 132
26  libbase.dylib                       0x00000001108769a5 non-virtual thunk to base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWork() + 21
27  libbase.dylib                       0x0000000110901df8 base::MessagePumpCFRunLoopBase::RunWork() + 200
28  libbase.dylib                       0x00000001108f9ac2 base::apple::CallWithEHFrame(void () block_pointer) + 10
29  libbase.dylib                       0x000000011090124f base::MessagePumpCFRunLoopBase::RunWorkSource(void*) + 63
30  CoreFoundation                      0x00007ff804473337 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
```
and still I'm only seeing infinite spinner after enabling brave news.
It's interesting that this happens only for local build.
Other official build works well. and feed list is empty while spinning like below.
Tried to add external source but it doesn't work.
I think I couldn't test brave news PR now. Sorry. can you find another reviewer?
image

File: components/brave_news/browser/feed_v2_builder.cc [hunk 2]
```diff
@@ -902,6 +904,45 @@ void FeedV2Builder::GenerateFeed(
...
+      GetArticleInfos(publishers_controller_->GetLastLocale(), feed_items,
+                      publishers_controller_->GetLastPublishers(), signals_);
+
+  auto feed = mojom::FeedV2::New();
+
+  constexpr size_t kIterationsPerAd = 2;
+  size_t blocks = 0;
+  while (!articles.empty()) {
+    std::vector<mojom::FeedItemV2Ptr> items;
+    items = GenerateBlock(articles, /*inline_discovery_ratio=*/0);
```
Comment (sangwoo108) [inline components/brave_news/browser/feed_v2_builder.cc:921]: Can we merge two lines into one?

File: components/brave_news/browser/feed_v2_builder.cc [hunk 3]
```diff
@@ -672,16 +680,10 @@ void FeedV2Builder::BuildPublisherFeed(const std::string& publisher_id,
                 continue;
               }
 
-              result->items.push_back(
-                  mojom::FeedItemV2::NewArticle(item->get_article()->Clone()));
+              items.push_back(item->Clone());
             }
 
-            base::ranges::sort(result->items, [](const auto& a, const auto& b) {
```
Comment (sangwoo108) [inline components/brave_news/browser/feed_v2_builder.cc:679]: Just double checking, we don't need sorting based on time and pop recency any more?

**Review by sangwoo108 — APPROVED** [review]
LGTM 👍

File: components/brave_news/browser/resources/feed/Article.tsx [hunk 4]
```diff
@@ -22,21 +22,15 @@ const Container = styled(Card)`
...
   const { url: imageUrl, setElementRef } = useLazyUnpaddedImageUrl(info.data.image.paddedImageUrl?.url ?? info.data.image.imageUrl?.url, { useCache: true })
   const url = info.data.url.url;
 
-  return <Container ref={setElementRef} onClick={() => openArticle(info.data)}>
+  return <Container ref={setElementRef} onClick={linkClickHandler(url)}>
     <ArticleMetaRow article={info.data} hideChannel={hideChannel} />
     <Flex direction='row' gap={spacing.m} justify='space-between'>
       <Title>
-        <SecureLink href={url}>{info.data.title}{('isDiscover' in info && info.isDiscover) && " (discovering)"}</SecureLink>
+        <SecureLink onClick={e => e.stopPropagation()} href={url}>{info.data.title}{('isDiscover' in info && info.isDiscover) && " (discovering)"}</SecureLink>
```
[the remark this answers is not in the document]
Comment (fallaciousreasoning) [reply components/brave_news/browser/resources/feed/Article.tsx:33]: The thing is, technically `a` tags shouldn't have interactive elements inside them, and there are some subtly weird things that happen when you do (I think there was some click/focus weirdness).

Additionally, nesting links doesn't work well at all :cry:

File: components/brave_news/browser/feed_v2_builder.cc (hunk 3 above)
Comment (fallaciousreasoning) [reply components/brave_news/browser/feed_v2_builder.cc:608]: I had a chat with lorenzo about it and I think it's fine. Channels should use exactly the same strategy for sorting as the main feed, and we're not sure what to do with publishers yet. In some cases, just sorting by recency makes the most sense, but in others, pop_score is more handy.

**Pull request merged**